### ISSUE TITLE
IsEmpty/IsDefault and Related Field Changes

### ISSUE THREAD

**Issue by robloo** [AvaloniaUI/Avalonia#7762]
**Describe the solution you'd like**

As discussed heavily in the `HsvColor` PR: https://github.com/AvaloniaUI/Avalonia/pull/7724#discussion_r817626579

`HsvColor` and `Color` structs should (IMO) have an `IsEmpty` property and an `Empty` field value (as named in System.Drawing.Color). This could also be named `IsDefault` and `Default` like in Xamarin.Forms.

These properties align these structures with older APIs that support this and indicate whether or not a Color struct is considered null or whether the system should interpret it with a transparent/default color instead. It doesn't need to be marked nullable with the added overhead that entails -- null can be communicated with the struct itself.

This idea is contested and I'm in the minority based on the linked discussion. Therefore, I've removed code that implements this from the `HsvColor` PR and created this discussion to make a final determination with a wider discussion.

---

This is still current in .NET. I didn't check other frameworks.

https://github.com/dotnet/runtime/blob/main/src/libraries/System.Drawing.Primitives/src/System/Drawing/Color.cs

Note that `IsEmpty` is tracked with a separate "state" field. That would mitigate any concern about it being equal to #00000000 which is technically a valid color.

Here are some .NET examples of `IsEmpty`:

 * [Color.Empty Field](https://docs.microsoft.com/en-us/dotnet/api/system.drawing.color.empty?view=net-6.0)
 * [Syncfusion IColor.IsEmpty](https://help.syncfusion.com/cr/file-formats/Syncfusion.Presentation.IColor.html#Syncfusion_Presentation_IColor_IsEmpty)
 * [GrapeCity Empty Field (Color)](https://www.grapecity.com/activereportsnet/docs/v14/online/GrapeCity.ActiveReports.Viewer.Wpf~GrapeCity.Viewer.Common.Color~Empty.html)

It was even done for other types:

 * [Point.IsEmpty Property](https://docs.microsoft.com/en-us/dotnet/api/system.drawing.point.isempty?view=net-6.0)
 * [Rectangle.Empty Field](https://docs.microsoft.com/en-us/dotnet/api/system.drawing.rectangle.empty?view=net-6.0)

And here is a .NET example of `IsDefault` with Xamarin.Forms:

 * [Color.IsDefault Property](https://docs.microsoft.com/en-us/dotnet/api/xamarin.forms.color.isdefault?view=xamarin-forms)

However, see the related discussion for why this was removed with MAUI: https://github.com/dotnet/Microsoft.Maui.Graphics/issues/57

**Describe alternatives you've considered**

The alternative is to add no Property/Field and just use a `nullable<Color>` when needed.

**Additional context**

**Comment by robloo** [comment]
Note that Avalonia has some inconsistency here at already exists:

 * Size has IsDefault
 * CornerRadius has IsEmpty
 * etc.

**Comment by jmacato** [comment]
@robloo would be great to harmonize those before 11 comes out

**Comment by robloo** [comment]
@jmacato I fully agree. But there was a lot of push-back by suggesting I add these properties with colors. My assumption was even these on Size/CornerRadius are considered unnecessary by some. Even though I know they are very useful.

There are also two cases we are talking about that people think about differently:
 1. IsEmpty when a struct was constructed without parameters so all properties are unset. This in Color/HsvColor is actually tracked with a separate state `int` that can differentiate between never-set and all components set-to-zero. (our discussion for HsvColor)
 2. IsEmpty when all components are set-to-zero. There is no distinction between uninitialized/null equivalent and just zero components. This is what is done in Size, CornerRadius, Rect, etc.

Once that distintion is worked out I will also need some direction on which name to use: IsEmpty/IsDefault (@Gillibald did suggest IsDefault). IsDefault may make more sense in all cases. However, then we should also consider the `Empty` field in some objects. We don't want 'Empty' and `IsDefault` together (which actually exists now in Size....) -- it should just be `Default` and `IsDefault` -- but Empty does sound better in some cases such as field usage.

Bottom line I'm open to suggestions. I haven't made any firm decisions yet and I don't think this discussion has ever been had before publicly in API design. So I'm not sure the original thinking back in WPF time-frame.

**Comment by robloo** [comment]
I think this is going to be needed for the ColorPicker no matter what. The new ColorPicker implementation will be very minimal code-behind and will have no way to distinguish what causes a color to change. Because of this, supporting a nullable/non-nullable Color property really isn't feasible since the nullable property would never know if the color was set as a default or by the user.

**Comment by robloo** [comment]
I've gone through this in some detail because I really think this part of the API needs to be cleaned up for 11.0 or it will never get fixed. I've summarized the main points below and included a table of some quick notes going through the code.

@jmacato @maxkatz6 @grokys I need approval/feedback from the core maintainers before going forward with any changes in this case.

The issue: As stated above, there are really 4 properties/fields to consider here `IsEmpty` / `Empty` and `IsDefault` / `Default`. These are not used in any standardized way. WPF was also not followed from what I can tell. Therefore, it was just left up to whomever was implementing the corresponding type to decide what to name these members and little thought was given at the time. We need to standardize convention and API here.

Before going into more detail, I looked through the code and summarized the following.

![image](https://user-images.githubusercontent.com/17993847/198842731-25287f75-1fbe-4435-acef-f61977caf988.png)

 * The critical and most-relevant types are highlighted with a green background (CornerRadius, Thickness, Point, etc.)
 * "None in WPF" means none of the 4 members IsEmpty, Empty, etc... exist in WPF
 * "Type not in WPF" is self explanatory
 * Note that this shows Avalonia follows no real convention and in some cases (Thickness) uses both

---

**So how to clean all of this up? I have the following ideas:**

 1. `IsDefault` / `Default` seems to be the "modern" convention and we should follow this for all types
     * Importantly, it matches with the `default` keyword in C#
     * "Empty" implies the type can have something in it. This makes sense for `string` or a collection but it doesn't make sense in most other cases. In common language, what does "empty thickness" mean? It can't hold anything. "default thickness" is much better.
 2. Compatibility with legacy code must be considered as well. In this case, compatibility with WPF not previous Avalonia code. For this case `Rect` is the only type that matches with WPF members of `IsEmpty` and `Empty`. It is probably a good idea to keep these around for 11.0 only with an Obsolete attribute and remove them in the next major release.
 3. Aside from `Rect` compatibility, no types will ever implement both `IsEmpty` / `IsDefault`. This means `Thickness` is going to loose the `IsEmpty` property.
 4. The `Default` field (like `string.Empty`) is meant to share a common instance for read-only types. This should be added to ALL relevant types even those that don't currently have it. So everything is going to have an `IsDefault` and `Default` going forward.
     * It needs to be decided if this should be a field or property though. Static read-only fields seem to be a bit more common in this case though.

After these changes, in a separate PR, I do intend to add IsDefault/Default back to the color structs. It really is useful in some cases to avoid the overhead of nullability.

**What are the core maintainer thoughts with this?**

**Comment by kekekeks** [comment]
>Compatibility with legacy code must be considered as well. In this case, compatibility with WPF not previous Avalonia code.

With 1.5M+ downloads, we should always be considering compatibility with previous versions. It shouldn't be broken unless absolutely necessary. We can mark old members as [Obsolete] though.

**Comment by maxkatz6** [comment]
I don't see a good reasoning in having `ValueType.Default` or `ValueType.Empty` types, when it can be replaced with `Rect value = default`. It makes more sense in checking if value is empty or not, like `value == Rect.Default`, but it can be replaced with `value.IsDefault`...

Also, I still don't see why IsDefault is better than IsEmpty. And same about the opposite.
Existing inconsistency doesn't make this problem any better though.

**Comment by robloo** [comment]
>>Compatibility with legacy code must be considered as well. In this case, compatibility with WPF not previous Avalonia code.
>
>With 1.5M+ downloads, we should always be considering compatibility with previous versions. It shouldn't be broken unless absolutely necessary. We can mark old members as [Obsolete] though.

True, I was thinking from the standpoint (1) Avalonia is alrady making breaking changes for 11.0 and (2) Obsolete members are being removed in this release already. This is a "clean" API to build from in the future. However, if you really want to maintain full compatibility with Avalonia until the next major release I can mark obsolete.

> I don't see a good reasoning in having ValueType.Default or ValueType.Empty types, when it can be replaced with Rect value = default. It makes more sense in checking if value is empty or not, like value == Rect.Default, but it can be replaced with value.IsDefault...

That makes sense and is a good point. Value types should not get the `.Default` member and we should use what C# already provides. Will update the spec for that.

> Also, I still don't see why IsDefault is better than IsEmpty. And same about the opposite.

 1. `IsDefault` matches well with the existing `default` keyword in C#
 2. "Empty" implies the type can have something in it. This makes sense for string or a collection but it doesn't make sense in most other cases. In common language, what does "empty thickness" mean? It can't hold anything. "default thickness" is much better.

> Existing inconsistency doesn't make this problem any better though.

Yes, so I'm trying to fix it

**Comment by robloo** [comment]
Unless there are objections I will plan to complete this within the next week or so. I will try to just obsolete those properties and fields which do not align with the new API.

**Comment by robloo** [comment]
Ran into a problem here. It seems the 'default' keyword will always create a new instance instead of sharing one i.e. `.Empty`. Therefore, the `Empty` instance property in principle probably needs to stay but renamed to `Default`... Let me know if there are comments/concerns for this.

**Comment by maxkatz6** [comment]
@robloo is it really a problem with value types? There should be somewhat zero overhead from creating a new default zeroed instance.

**Comment by robloo** [comment]
> @robloo is it really a problem with value types? There should be somewhat zero overhead from creating a new default zeroed instance.

Yea, you're right. Of course value types are "copied" anyway. I wasn't thinking specific enough. The only two places this applies are `Rect` and `PixelRect` which are both readonly structs.

**Comment by maxkatz6** [comment]
TIL, WPF Rect.IsEmpty is not the same as IsDefault
See https://github.com/dotnet/wpf/blob/4dbc072e3f2c30930a7a9ff9dab9114660ae2b88/src/Microsoft.DotNet.Wpf/src/WindowsBase/System/Windows/Rect.cs#L134-L148

**Comment by robloo** [comment]
Yes, IsEmpty has a few different nuances. In the older Color types it also doesn't mean #00000000 but is actually zero plus an additional "unset" state flag.

In this case I think the IsDefault standardization simplified things a bit overall. Avalonia was differing for Rect it seems anyway and was using zeros for the default case.

In the future we could always add an IsZero property to be more clear and differentiate from IsDefault which may not have zero in all cases. I still think IsEmpty is a bit of a misnomer for anything other than collections.
